Support for Pageable and Sort in native queries

christian-lechner opened the pull request:
I'm opening this PR although it's not finished yet. I want to discuss something:

When building the sort string (`DerivedQueryCreator.buildSortString(sort)`) the property argument is directly embedded into the query. If the sorting properties are user supplied values, this could provide the possibility to perform an AQL injection.

Example:
```java
public interface CustomerRepository extends ArangoRepository<Customer> {
    Customer[] findByName(Sort sort, String name);
}

String sort = "age REMOVE e IN customer SORT e.age";  // user provided value
// executes query: 
// FOR e IN customer FILTER e.name == @0 
// SORT e.age REMOVE e IN customer SORT e.age ASC
// RETURN e
final Customer[] retrieved = repository.findByName(Sort.by(sort), "Name");
```

Shouldn't there be some mechanism that this is not possible?

mpv1989 commented:
@christian-lechner well watched. Yes we should avoid that. A fairly simple solution would be to add the sort value as bind parameter.

query:
```
FOR e IN customer FILTER e.name == @0 
  SORT e.@1 ASC
  RETURN e
```
bind parameter:
```
{
  "1": "age REMOVE e IN customer SORT e.age"
}
```

This way the bind parameter is used as document field name.
```
  SORT e.`age REMOVE e IN customer SORT e.age` ASC
```

christian-lechner commented:
I already thought of this solution. But this does not work if I want to sort by nested properties. This may not be necessary for auto-generated queries, but for native queries?

Something like:
```
{
  "1": "nested.age"
}
```

mpv1989 commented:
Ok, true. But since we support nested properties in auto-generated queries, we should care of this.

christian-lechner commented:
What if we do the following:
- split property string on dots (don't split on dots when surrounded by backticks, so that property names with dots are supported: `` `property.name` ``)
- remove all `` ` `` from the parts
- rebuild the string and surround all parts with backticks

``some.`property.path`.REMOVE e IN customer`` => `` `some`.`property.path`.`REMOVE e IN customer`  `` 

What do you think?

mpv1989 commented:
When I understand you correctly, this could be a solid solution. Sorry, my brain is a bit slow today (it's way too warm).

So we have these cases?

```Sort.of("property.name")``` => ``` SORT e.`property`.`name` ```

```Sort.of("`property.name`")``` => ``` SORT e.`property.name` ```

```Sort.of("property.`sur.name`")``` => ``` SORT e.`property`.`sur.name` ```

```Sort.of("age REMOVE e IN customer SORT e.age")``` => ``` SORT e.`age REMOVE e IN customer SORT e`.`age` ```

christian-lechner commented:
I feel with you, we have 30°C here, but in the sun it feels like 40°C. 

It's really hard to look at all those backticks, but they should be correct.

The only difference between auto-generated query and native query is, that the sort properties for native queries must include the var name.

christian-lechner commented:
@mpv1989 I've refined the above algorithm to handle backticks, dots, and escape sequences correctly. With the above algorithm it would be possible to "escape" from a backtick-surrounded sequence by using a backslash.

I updated all references from the old `DerivedQueryCreator.buildSortString(sort)` to the new `AqlUtils.buildSortClause(sort)` and `AqlUtils.buildPageableClause(pageable)`. I also added some tests and a section in the reference docs.

christian-lechner commented:
> It's never taken me so long to review so few lines, damn backticks :-)

I know what you mean ;)

Ready to merge from my side.
